kanidm: init at 1.1.0-alpha.8, add module

erictapen opened the pull request:
[Kanidm](https://kanidm.github.io/kanidm/) is an identity management platform written in Rust, that I'm very excited about personally.

I only added module suport for the web application. Kanidm also comes with an optional daemon for PAM/nsswitch integration, but I didn't use it yet and therefore couldn't test it.

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Flakebi commented:
Hi, it’s nice too see a pull-request for kanidm. I’m using it for quite a while but never got it clean enough to submit it.
I’m using the PAM integration and – since last week – also the oidc support to connect it to matrix-synapse and hedgedoc.
Maybe you can take some inspiration from the module: https://github.com/Flakebi/nixpkgs/commit/085df2dbebad0f3ceff470f28c30f116ea4a5e10

erictapen commented:
@Flakebi awesome to see that you came up with an alternative module. Also nice that you got the PAM integration set up. My main use case for kanidm currently is to use it for OIDC, so I didn't really care about PAM, but having it in the NixOS module would be great.

We could work on this together? You could rebase your branch ontop of my on in order to add PAM integration? Then I could pull your commits in. From a first glance you PAM integration looks pretty fine. Only it would probably make sense to also test it in the NixOS integration test.

Flakebi commented on nixos/modules/services/databases/kanidm.nix:
```diff
@@ -0,0 +1,101 @@
+        after = [ "network.target" ];
+        wantedBy = [ "multi-user.target" ];
+        serviceConfig = {
+          ExecPreStart = ''
+            ${pkgs.kanidm}/bin/kanidmd domain_name_change \
+              --config ${settingsFormat.generate "config.toml" cfg.settings} \
+              -n ${cfg.ensureDomainName}
+          '';
```
This seems to be a pretty expensive operation and according to the kanidm docs
> WARNING: This WILL break ALL u2f/webauthn tokens that have been enrolled, which MAY cause accounts to be locked out and unrecoverable until further action is taken. DO NOT CHANGE the domain_name unless REQUIRED and have a plan on how to manage these issues.

So, I think if we add support to change the domain name in nix, we only want to run this if the domain is different.

erictapen replied:
I assumed naively, that it doesn't have a cost at all if the name isn't changed? At least thats what I'd expect from the implementation. I didn't find a way to fetch the current domain_name from the db. If there is one we could introduce a check.

Flakebi replied:
I briefly looked at the implementation but I didn’t see a check to do nothing if there is no change. We could store the domain in a text file near the db file and only run this if it changed.

erictapen replied:
Maybe we could ask upstream to implement either such a check or a possibilty to retrieve the domain name by querying the domain? Adding a text file would work, but I'm somewhat hesitant to add it as a de-facto config file, considering that upstream explicitly doesn't want to have `domain_name` as config, but as a database property:

https://github.com/kanidm/kanidm/issues/616
https://github.com/kanidm/kanidm/pull/617#issuecomment-978350062

Do you agree in principle that the `ensureDomainName` should be a thing?

Flakebi replied:
I’d like it if things get more declarative. Both of your ideas sound better than having a text file :)

mweinelt reviewed (commented):
There seem to be two competing implementations and it probably makes sense to take the best of both worlds.

- This implementation seems to have support for domain name migrations and considers the different synchronization roles.

- Flakebis module implementation support multiple daemons and has PAM/NSSwitch integration. Also I very much like the hardening done in that module.

- Also there is at least a third implementation by @andir at https://git.rammhold.de/kanidm-nix/tree/ that can be used for inspiration.

Build-tested on both aarch64-linux and x86-64.

Would be sad to see this go stale. Kanidm looks very promising and I'm looking forward to using it for OIDC and RADIUS in the near future.

Flakebi commented:
@mweinelt, I just updated https://github.com/erictapen/nixpkgs/pull/1, which is @erictapen’s package and their module improvements on top of my module.

Flakebi commented on nixos/modules/module-list.nix:
```diff
@@ -333,6 +333,7 @@
   ./services/databases/hbase.nix
   ./services/databases/influxdb.nix
   ./services/databases/influxdb2.nix
+  ./services/databases/kanidm.nix
```
pam is in `modules/security` and openldap is in `modules/databases`, I think both directories are fitting.

mweinelt replied:
Kanidm describes itself as an IDM, while I know that in OpenLDAP your interaction is more database-like, with schemas, freeform tree structures, etc., hence the thought that they are differently scoped.

Flakebi replied:
True, I moved it in my PR

mweinelt commented on nixos/modules/services/security/kanidm.nix:
```diff
@@ -0,0 +1,335 @@
+
+    ensureDomainName = lib.mkOption {
+      description = ''
+        The domain_name that Kanidm manages. It must be a descendant or equal to the domain
+        specified in settings.origin. It is always set when the server starts up.
+        While in theory it could be changed later, you should carefully consider these
+        warnings:
+        <link xlink:href="https://kanidm.github.io/kanidm/administrivia.html#rename-the-domain"/>
```
```suggestion
        The <literal>domain_name</literal> that Kanidm manages. Must be below or equal to the domain
        specified in <literal>serverSettings.origin</literal>. Will always be set when the server starts up.
        
        While it is possible to change this setting later, conider the warnings in the
        <link xlink:href="https://kanidm.github.io/kanidm/administrivia.html#rename-the-domain">relevant documentation</link>.
```

mweinelt commented on nixos/modules/services/security/kanidm.nix:
```diff
@@ -0,0 +1,335 @@
+            type = lib.types.str;
+          };
+          ldapbindaddress = lib.mkOption {
+            description = "Address/port combination the emulated ldap server binds to.";
```
```suggestion
            description = ''
              Address and port the LDAP server is bound to. Setting this to an empty string disables the LDAP interface.
            '';
```

Better explains the behaviour of the default value.

mweinelt reviewed (changes requested):
Did a first pass over the module. Some option descriptions need an updated after the module merge. Kanidm is a name and should be capitalized when named in descriptions. 

Did not look at the tests yet. Also the CI is complaining.

Flakebi commented on nixos/modules/services/security/kanidm.nix:
```diff
@@ -0,0 +1,335 @@
+        StateDirectory = "kanidm";
+        StateDirectoryMode = "0700";
+        ExecStart = "${cfg.package}/bin/kanidmd server -c ${serverConfigFile}";
+        ExecPreStart = ''
+          ${cfg.package}/bin/kanidmd domain_name_change \
+            -c ${serverConfigFile} \
+            -n ${cfg.ensureDomainName}
+        '';
```
That has been resolved upstream, so `domain_name_change` is now cheap if it doesn’t actually change.

mweinelt commented:
Please set `meta.buildDocsInSandbox = false` in the module to fix the manual build.

erictapen commented on nixos/modules/services/security/kanidm.nix:
```diff
@@ -0,0 +1,335 @@
+          };
+          db_path = lib.mkOption {
+            description = "Path to kanidm database.";
+            default = "/var/lib/kanidm/kanidm.db";
```
I agree here. Made it readonly in 876618d0451cb93c089fb52b1c1a6f226099d3e2.

Not sure I understand your ZFS setup, but if your bind mounts are managed by systemd mounts, they are dependencies automatically! By setting `StateDirectory = "kanidm"` systemd adds `RequiresMountsFor=/var/lib/kanidm` to the unit. This way I'm able to have most of my directories in `/var/lib` as ZFS datasets without extra config.

erictapen commented on nixos/modules/services/security/kanidm.nix:
```diff
@@ -0,0 +1,335 @@
+          origin = lib.mkOption {
+            description = "The origin of your kanidm instance. Must have https as protocol.";
+            example = "https://idm.example.org";
+            type = lib.types.str;
```
Nice, didn't know about that!

erictapen commented:
@Flakebi I'd like to remove the `package` option, as I did in 936819a10770e82678e271beb768438fb3cbd960.

The general consensus seems to me that a package option is only added if there is an alternative implementation available in Nixpkgs. Otherwise it's cleaner to override the package in an overlay imho.

erictapen commented:
@mweinelt I'd rather leave backup out of the scope for this initial PR, so this can be used by many people soon.

The one priority I currently have left is improving the test: I'd like to test the kanidm client interaction with the server and the PAM/NSS integration. Testing LDAP integration and maybe even OIDC would be nice too.

mweinelt commented on nixos/tests/kanidm.nix:
```diff
@@ -0,0 +1,58 @@
+    nodes.client = { pkgs, nodes, ... }: {
+      environment.systemPackages = [ pkgs.kanidm ];
+
+      environment.etc."kanidm/config".text = ''
+        uri = "https://${serverDomain}"
+        verify_ca = true
+        verify_hostnames = true
+      '';
```
Needs to be migrated to the module based configuration.

mweinelt commented on nixos/tests/kanidm.nix:
```diff
@@ -0,0 +1,58 @@
+    admin = {
+      username = "admin";
+      password = "snakeoilpass";
+    };
```
This looks unused.

mweinelt commented on nixos/modules/services/security/kanidm.nix:
```diff
@@ -0,0 +1,335 @@
+        StateDirectory = "kanidm";
+        StateDirectoryMode = "0700";
+        ExecStart = "${pkgs.kanidm}/bin/kanidmd server -c ${serverConfigFile}";
+        ExecPreStart = ''
```
```suggestion
        ExecStartPre = ''
```

mweinelt commented on nixos/modules/services/security/kanidm.nix:
```diff
@@ -0,0 +1,335 @@
+        ExecPreStart = ''
+          ${pkgs.kanidm}/bin/kanidmd domain_name_change \
+            -c ${serverConfigFile} \
+            -n ${cfg.ensureDomainName}
```
This is still a bit iffy. I ran into the problem that example.com was preconfigured and domain change would fail while my origin wasn't below example.com.

mweinelt commented on nixos/modules/services/security/kanidm.nix:
```diff
@@ -0,0 +1,335 @@
+        Group = "kanidm";
+
+        AmbientCapabilities = "CAP_NET_BIND_SERVICE";
+        CapabilityBoundingSet = "CAP_NET_BIND_SERVICE";
```
> Could not bind to ldap server address [::1]:636 -> 13, kind: PermissionDenied, message: "Permission denied" }

Not sure why yet.

Firstyear replied:
Because ports below 1024 are protected

mweinelt replied:
Yeah, but that's why we give the service the `CAP_NET_BIND_SERVICE` capability.

>        CAP_NET_BIND_SERVICE
>              Bind a socket to Internet domain privileged ports (port
>              numbers less than 1024).

https://man7.org/linux/man-pages/man7/capabilities.7.html

erictapen replied:
`PrivateUsers=true` was the culprit here, as it creates a user namespace and applies the CAP_NET_BIND_SERVICE only in that namespace. I disabled it for the service and introduced testing for the open ldap port in 46345ff70ed9e041d9f1c0cbd2edcddece4f5f83.

mweinelt commented:
> I only added module suport for the web application.

But we didn't package the webui AFAIU, is that correct?

```
 INFO     ┝━ 💬 [request.info]: Request received | src: "fd42:23:42:b864:7285:c2ff:fe67:b6c9" | http.host: "idm.lossy.network" | http.method: "GET" | path: "/pkg/bundle.js"
 WARN     ┕━ 🚧 [request.warn]: Client error --> Response sent | status: 404 - Not Found
```

Flakebi commented:
@mweinelt thanks for the review, I put up some fixes in https://github.com/erictapen/nixpkgs/pull/2
That should fix the domain name change, the web ui and pam/nss libraries.

erictapen commented on nixos/modules/services/security/kanidm.nix:
```diff
@@ -0,0 +1,335 @@
+              Address and port the LDAP server is bound to. Setting this to an empty string disables the LDAP interface.
+            '';
+            example = "[::1]:636";
+            default = "";
```
Fixed in 964ec52d954a7d3f666769ae8e32b29223dad550!

mweinelt replied:
I agree that filtering out null values is a better approach than removing options. CC @Flakebi

erictapen commented:
After thinking about it for a while I'd like to change the structure of the module. Imho kanidm should be configured as such:

```nix
{ ... }: {

  services.kanidm = {
    enable = true;
    ensureDomainName = "example.org";
    settings = { ... };                 # currently serverSettings
    pam = { # or maybe call it unix?
      enable = true;
      settings = { .. };                # currently unixSettings
    };
  };

  programs.kanidm = {
    enable = true;
    settings = { .. };                  # currently clientSettings
  };

}
```

That would require moving part of the logic into `nixos/modules/programs/kanidm.nix`.

What do you think about that? I'd start with the restructuring after https://github.com/erictapen/nixpkgs/pull/2 is merged.

mweinelt commented:
Actually instead of `programs.kanidm` it should likely be `security.kanidm`, wouldn't you agree?

erictapen commented:
> Actually instead of `programs.kanidm` it should likely be `security.kanidm`, wouldn't you agree?

I don't think so, as `kanidm` strikes me purely as a client program. It doesn't interact with the operating system at all, only with the remote service. I always interpreted `security.*` as modules with deep OS integration, e.g. polkit or acme.

One similar case would be openssh: `programs.ssh` is for client config, `services.openssh` for the server.

Flakebi commented:
Nice fix for the ldap option!

> just realised that the PAM daemon also uses the `/etc/kanidm/config` file. Now I get your point.

Yeah, that was my reason to leave everything under `services.kanidm`. I find it easier to discover if everything is under one scope (e.g. through https://search.nixos.org/options) and if one doesn’t need to touch `programs.kanidm` to configure the pam service in `services.kanidm`.

From what I can see, the `security.pam` options _only_ change the pam configuration and do not enable any services. Instead, the dovecot service has an `enablePAM` option that also sets `security.pam.services.dovecot2` (cups and doas seem to do something similar). I guess it makes sense to follow the same scheme for kanidm.
I have a working `config.security.pam` option locally, however, I needed some changes in the pam config that probably shouldn’t be upstreamed, so I need to write a test and trim down the diff before it’s ready. It may take a while until I find the time and motivation to do that, so I don’t want to stall this PR for it :)

mweinelt commented:
The upstream did a refactor of the domain name change handling. Can we upgrade to an unstable version and update our prestart script as needed? It should hopefully get become simpler.

https://github.com/kanidm/kanidm/pull/639

Firstyear commented:
@mweinelt We also have a release on april 1st if you can wait for a month or so :)

mweinelt commented:
This goes into the unstable branch and our next release is at the end of may, so I think we can base our work off of an unstable version and upgrade to the next stable release before the branch-off.

mweinelt commented:
Hey, how is everbody? This has been a bit quieter than I had hoped for. Do people still have the energy to finish this?

Firstyear commented:
> Hey, how is everbody? This has been a bit quieter than I had hoped for. Do people still have the energy to finish this?

Next release is may 1st which has the fixes here. We slid the schedule by 1 month to allow for some other things to be developed, and to prevent releasing on jan 1st each year. 

@Flakebi @erictapen I'll communicate with you both when the next release is ready so this can go ahead?

Flakebi commented:
I tried updating to the master branch a few weeks ago, but kanidm_domain_change still complained about the wrong origin (or domain?) and I didn’t have time to investigate more since then.

erictapen commented:
I think the new kanidm release will give me the push to work some more on this :)

erictapen commented:
Yeah, the notifications are enough, thanks!

erictapen commented:
- Bumped to the 1.1.0-alpha.8 release
- removed our `ensureDomainName` logic and replaced it with `serverSettings.domain`, accompanied with references to the docs
- rebased on latest master, as we need the newer rustc

mweinelt commented on nixos/modules/services/security/kanidm.nix:
```diff
@@ -0,0 +1,343 @@
+        }
+        {
+          assertion = !cfg.enableServer || (isNull cfg.serverSettings.domain
+            -> cfg.settings.role == "WriteReplica" || cfg.settings.role == "WriteReplicaNoUI");
```
```suggestion
            -> cfg.serverSettings.role == "WriteReplica" || cfg.serverSettings.role == "WriteReplicaNoUI");
```

erictapen replied:
Good catch! 957e2d6ca40f4c7edd314209a3e36b99e72fb59c

mweinelt commented:
I tested this previously on a production host with 21.11, but due to the latest glibc 2.34 bump on unstable that doesn't work anymore. It will take me a while to migrate that setup elsewhere.

erictapen commented:
@mweinelt Interesting, I'm doing the same but the service runs fine for me? How do you compose unstable and stable? I'm just overriding kanidm in an overlay and I add the module to `extraModules`.

mweinelt commented:
I copied both module and package into my local config tree, substituted `cfg.package` for `package` which references `pkgs.callPackage ./package.nix` and migrated `rustPlatform` to its nixos-unstable counterpart.

mweinelt commented:
Does recovering the admin account work for you? It errors out for me and I'm not sure what account to login with there.

```
# kanidm account credential reset_credential  --name admin idm_admin
2022-05-01T14:36:05.510597Z ERROR kanidm_cli::common: No valid authentication tokens found. Please login with the 'login' subcommand.

# kanidm login --name anonymous
Login Success for anonymous

# kanidm account credential reset_credential  --name admin idm_admin
2022-05-01T14:43:34.840094Z ERROR kanidm_cli::common: No valid authentication tokens found for admin.
```

I think it's a basic task that we should include in the test.

erictapen commented:
@mweinelt I think you have to login as admin user first. Set an admin password like this: https://kanidm.github.io/kanidm/stable/server_configuration.html#default-admin-account

I definetly want to include that in the test, but retrieving the password from the output might get cumbersome.

erictapen commented:
From my point of view, there aren't any outstanding issues left, so to me it looks like it's good to merge.
Test coverage could be greatly improved for sure, but I'd like to keep that out of the scope of this PR. It's already better than many tests in tree, that just wait for the systemd unit.

What do you think, y'all? If anybody is willing to merge this, I'll squash the PR into two commits (one for package, one for the module).

mweinelt commented:
```
# kanidmd recover_account -c /nix/store/nbcbk7a82yb6mn9jl7iwq6ziqh09rrrs-server.toml -n admin |& grep "password reset to" | awk '{ print $NF }'
1uffaUP9cQUKFgEpcZfqAr83EpyD7JsMvHMD0zGu6s3vhXc6
```

The actual question would be how to get the config path.

erictapen commented:
I already have something in the works for grabbing the generated passwort, but unfortunately I'm having trouble passing it to `kanidm login`. Any ideas?

Flakebi reviewed (approved):
I tested the updated and it works fine for me.

mweinelt commented:
> I already have something in the works for grabbing the generated passwort, but unfortunately I'm having trouble passing it to `kanidm login`. Any ideas?

A small expect script would do this.

mweinelt commented on nixos/modules/services/security/kanidm.nix:
```diff
@@ -0,0 +1,345 @@
+          '';
+        }
+        {
+          assertion = !cfg.enableServer || (isNull cfg.serverSettings.domain
```
```suggestion
          assertion = !cfg.enableServer || (cfg.serverSettings.domain == null
```

https://nixos.org/manual/nix/stable/expressions/builtins.html#builtins-isNull

erictapen replied:
:+1:

mweinelt commented on pkgs/servers/kanidm/default.nix:
```diff
@@ -0,0 +1,89 @@
+, nixosTests
+, rustPlatform
+, fetchFromGitHub
+, runCommand
```
unused

erictapen replied:
:+1:

mweinelt reviewed (commented):
Two nits and a request for release notes.

erictapen commented:
@mweinelt Done!

erictapen commented:
@Artturin does the `12.approvals: 2` label mean this PR needs two approving reviews? Is there some sort of documentation for ths policy? Personally I'd expect this fine to merge with one approval, as it won't break anybody's setup.

Artturin commented:
> @Artturin does the `12.approvals: 2` label mean this PR needs two approving reviews? Is there some sort of documentation for ths policy? Personally I'd expect this fine to merge with one approval, as it won't break anybody's setup.

No, it is the count of how many people have approved this pr 
It was added automatically by the following scripts. you can see the reasoning for creating it in the readme
https://github.com/nix-community/label-approved

erictapen commented:
@Artturin I see, thanks!

@mweinelt Would you like to have this approved by somebody else or would it be ok for you if I self-merge? I really want to get this in-tree so it can be used easily by others.

mweinelt commented:
a) You wanted to squash these changes, I'm happy with merging it as si
b) I'm unclear if you wanted to look further into the password recovery and admin login flow or leave it at that.

erictapen committed:
kanidm: init at 1.1.0-alpha.8

Co-Authored-By:  Martin Weinelt <mweinelt@users.noreply.github.com>
Co-Authored-By:  Flakebi <flakebi@t-online.de>

erictapen committed:
nixos/kanidm: init

Co-Authored-By:  Martin Weinelt <mweinelt@users.noreply.github.com>
Co-Authored-By:  Flakebi <flakebi@t-online.de>

erictapen commented:
I'm sorry, completely forgot about sqashing here.
Let's leave the password login tests for a later PR, so feel free to merge :+1:

erictapen commented:
Thanks a lot to @Flakebi @mweinelt @Firstyear for your work on this! :partying_face:

Artturin commented:
there's a ci failure..

erictapen commented:
Damn, sorry. Fixed in https://github.com/NixOS/nixpkgs/pull/171688.
